### ISSUE TITLE
Consider simplifying the `ConstExpr`/`Expr` divide

### ISSUE THREAD

**Issue by mystor** [dtolnay/syn#171]
Right now if you turn on `full`, you get a super sub-optimal experience, where in some places you need a `ConstExpr` and in others you need an `Expr`, and it's not consistent. For example, if you're using the `Folder` from `fold`, and you want to perform a transformation on every expression, you need to register a `fold_const_expr` and `fold_expr` callback separately. In addition, some work, such as the precedence work which I've done as part of #168, doesn't affect `ConstExpr`s right now, as they go through a completely separate parsing path.

This isn't consistent with the rust language, which doesn't actually make the distinction. It's done in `syn` to save on compile time, as accepting arbitrary expressions means you have to parse the entire language, which takes significantly longer to compile.

One option would be to keep the exact same `Expr` AST with full enabled and disabled, with the disabled inner structs stubbed out as structs with private constructors. The parsing and printing code `cfg()`-ed out for these types. I think this would get us the compile time advantages, while not requiring the separate `ConstExpr` type.

To show what I mean, here's what `ExprKind` might look like:
```rust
// ExprKind would be unchanged
pub enum ExprKind {
    Binary(ExprBinary),
    If(ExprIf),
    // ...
}

// In full mode we would have everything as it is now:
#[cfg(feature = "full")]
pub struct ExprIf {
    pub cond: Box<Expr>,
    pub if_true: Block,
    pub if_false: Option<Box<Expr>>,
    pub if_token: tokens::If,
    pub else_token: Option<tokens::Else>,
}

// But in non-full mode, the types would be impossible to construct or obtain,
// and wouldn't have `ToTokens` or `Synom` implementations. (Or perhaps
// they would have a `Synom` implementation, which would always fail).
#[cfg(not(feature = "full"))]
pub struct ExprIf {
    _no_construct: ()
}
```

**Comment by alexcrichton** [comment]
Seems plausible to me!

**Comment by mystor** [comment]
Fixed by #175
